docs/developer/developing-plugins.md

```diff
@@ -0,0 +1,148 @@
+}
+
+func (s *Socket) Config(c []byte) error {
+    s.conf = configT{}
+    err := config.ParseConfig(bytes.NewReader(c), &s.conf)
+    if err != nil {
+        return err // sg-core will log configuration error message and exit
+    }
+    return nil
+}
+
+[...]
+
+func New(l *logging.Logger) transport.Transport {
+    return &Socket{
+        logger: l,
+        conf: configT{
+            Port: 9090 // set defaults. Only overriden if input data for Config() contains the port option
+        }
+    }
+}
+
+```
+
+## Transports
+
+Transport plugins listen on an external messaging protocal and deliver received messages to handlers that have been bound to it per the administrator's configuration. They receive a configuration block from the main configuration file and write to handlers by involing transport.WriteFn in Run().
+
+Transports should contain the minimal amount of code necessary to fulfill this functionality. 
+
+Transport plugin objects must implement the the Transport interface:
+```go
+type Transport interface {
+	Config([]byte) error
+	Run(context.Context, *sync.WaitGroup, transport.WriteFn, chan bool)
+	Listen(data.Event)
+}
+```
+
+## Handlers
+
+Handlers parse incoming blobs from the transport into objects and delivers those objects to the internal buses. There are two types of handlers: metric handlers and event handlers. Metric handlers deliver metric objects to the internal metrics bus while event handlers deliver event objects to the internal events bus. These metrics and events are then consumed by the application plugins.
+
+Handlers must remain simple: they take in no configuration and should only handle one message type. See the [collectd-metrics](https://github.com/infrawatch/sg-core/tree/master/plugins/handler/collectd-metrics) plugin for an example. Additionally, handlers should not print logs (this is why no logger is passed into the New() function). If errors occur while parsing messages, handlers should create their own metrics or events recording the error(s) and write them to the bus. The `collectd-metrics` handler iterates a counter every time a parsing error occurs and submits the number as a metric to the metric bus.
+
+
+Handler plugin objects must imeplement either the MetricHandler interface or the EventHandler interface:
+```go
+type MetricHandler interface {
+	Handle([]byte) []data.Metric
+}
+
+type EventHandler interface {
+	Handle([]byte) (data.Event, error)
+}
+```
+
+## Applications
+
+The purpose of application plugins are to provide the business logic for interfacing with external programs like a database. They receive both metrics and events and must decide what to do with them. For example, the [prometheus](https://github.com/infrawatch/sg-core/tree/master/plugins/application/prometheus) plugin receives metrics from the internal metrics bus and stores them into Prometheus.
+
+Application plugins must implement the Application interface:
+```go
+type Application interface {
+	Config([]byte) error
```

Comment:
Can the Config contain whether the apps wants a metric and/or an event bus?  The manager would then be able to supply the appropriate channel(s).

Reply:
Yeah that's a good idea. That way, the admin can choose which messages should be received by the application.